### PR TITLE
A benchmark suite for the `getAccountInfo` RPC call

### PR DESCRIPTION
**Pull request by steveluscher**
#### Summary of Changes

```shell
cd accounts-cluster-bench/
cargo run -- --identity ~/.config/solana/id.json --rpc-bench account-info --iterations 0 --url l

[2024-12-05T22:23:35.196026504Z INFO  solana_accounts_cluster_bench] creating 4 new
[2024-12-05T22:23:35.196792009Z INFO  solana_accounts_cluster_bench] txs: 4
[2024-12-05T22:23:35.202978709Z INFO  solana_accounts_cluster_bench] ids: 4
[2024-12-05T22:23:35.203003136Z INFO  solana_accounts_cluster_bench] total_accounts_created: 368 total_accounts_closed: 0 tx_sent_count: 368 loop_count: 103 balance(s): [732814816240]
[2024-12-05T22:23:35.426296399Z INFO  solana_accounts_cluster_bench] t(0) rpc(AccountInfo) iters: 67827 success: 5228 errors: 19
[2024-12-05T22:23:35.426334171Z INFO  solana_accounts_cluster_bench]  t(0) rpc(AccountInfo average success_time: 761 us
[2024-12-05T22:23:35.426344090Z INFO  solana_accounts_cluster_bench]  rpc average average errors time: 718 us
```

Related: #3242.

### CODE REVIEW

File: accounts-cluster-bench/src/main.rs [hunk 1]
```diff
@@ -340,6 +342,35 @@ fn run_rpc_bench_loop(
             break;
         }
         match rpc_bench {
+            RpcBench::GetAccountInfo => {
+                let mut rpc_time = Measure::start("rpc-get-account-info");
+                let start: u64 = max_closed.load(Ordering::Relaxed);
+                let end: u64 = max_created.load(Ordering::Relaxed);
+                let seed_range = start..end;
+                if seed_range.is_empty() {
+                    info!("get_account_info: No accounts have yet been created; skipping");
```
Comment (steveluscher) [inline accounts-cluster-bench/src/main.rs:400]: See note here: #3960.

File: accounts-cluster-bench/src/main.rs [hunk 2]
```diff
@@ -340,6 +342,35 @@ fn run_rpc_bench_loop(
...
+                    continue;
+                }
+                let seed = thread_rng().gen_range(seed_range).to_string();
+                let account_pubkey =
+                    Pubkey::create_with_seed(&base_keypair_pubkey, &seed, &program_id).unwrap();
+                match client.get_account(&account_pubkey) {
```
Comment (steveluscher) [inline accounts-cluster-bench/src/main.rs:407]: Some percentage of these loads fail with:

```shell
[2024-12-05T22:23:08.309451301Z INFO  solana_accounts_cluster_bench] get_account_info error: Error { request: None, kind: RpcError(ForUser("AccountNotFound: pubkey=Bp88sz54nzCwcxv8S7K6s6Yw5MX4H2FVwVm5sufvBDPa")) }
```

I don't know if `executor.pushTransactions()` is sufficient to presume that the accounts have _actually_ been created.

https://github.com/anza-xyz/agave/blob/master/accounts-cluster-bench/src/main.rs#L643

Is that what's going on here? Is it some sort of race, where I try to get the account before the transaction has committed?

File: accounts-cluster-bench/src/main.rs [hunk 3]
```diff
@@ -340,6 +342,35 @@ fn run_rpc_bench_loop(
...
+            RpcBench::GetAccountInfo => {
+                let mut rpc_time = Measure::start("rpc-get-account-info");
+                let start: u64 = max_closed.load(Ordering::Relaxed);
+                let end: u64 = max_created.load(Ordering::Relaxed);
+                let seed_range = start..end;
+                if seed_range.is_empty() {
+                    info!("get_account_info: No accounts have yet been created; skipping");
+                    continue;
+                }
+                let seed = thread_rng().gen_range(seed_range).to_string();
```
Comment (steveluscher) [inline accounts-cluster-bench/src/main.rs:403]: Samples randomly from the range of seeds belonging to open accounts.

File: accounts-cluster-bench/src/main.rs [hunk 4]
```diff
@@ -683,6 +720,8 @@ fn run_accounts_bench(
             let _ = executor.drain_cleared();
         }
 
+        start_bench_tx.send(()).unwrap();
```
Comment (alessandrod) [inline accounts-cluster-bench/src/main.rs]: won't this only unblock one of the bench threads?

For this kind of stuff, I think
https://doc.rust-lang.org/stable/std/sync/struct.Barrier.html would be more idiomatic

Comment (alessandrod) [reply accounts-cluster-bench/src/main.rs]: Ah, looks like stacking got me already 😅

File: accounts-cluster-bench/src/main.rs (hunk 2 above)
Comment (alessandrod) [reply accounts-cluster-bench/src/main.rs:407]: Correct, when push_transactions() returns the txs have been queued but not
guaranteed to have been processed and included in a block.

The code that create accounts is _very_ involved, not sure it's easy to change
to guarantee this.

**Review by alessandrod — DISMISSED** [review]
The bench here looks good. For the channel unblock thing, see the other PR.

File: accounts-cluster-bench/src/main.rs [hunk 5]
```diff
@@ -340,6 +343,35 @@ fn run_rpc_bench_loop(
             break;
         }
         match rpc_bench {
+            RpcBench::GetAccountInfo => {
+                let mut rpc_time = Measure::start("rpc-get-account-info");
```
Comment (t-nelson) [inline accounts-cluster-bench/src/main.rs]: why are we timing the setup?

Comment (steveluscher) [reply accounts-cluster-bench/src/main.rs]: We're just super dumb.

ty for the catch.

File: accounts-cluster-bench/src/main.rs (hunk 4 above)
Comment (steveluscher) [reply accounts-cluster-bench/src/main.rs]: > won't this only unblock one of the bench threads?

Dude, I don't know how computers work.

ty for the pointer.

Comment (steveluscher) [reply accounts-cluster-bench/src/main.rs]: @nickfrosty, this is – in my opinion – a perfect example of what we were talking about earlier today. In this particular instance I was on the ‘don't know what I don't know‘ side of the house. I needed a _guide_ on synchronization primitives. I spent a ton of time with the Rust API docs, but they didn't lead me to the right solution. Furthermore, as soon as Alessandro gave me the pointer to `Barrier`, I no longer had any _use_ for the web-based API docs; I found all the answers I needed in the Rust LSP / the code itself.

**Comment by steveluscher** [comment]
Updated to use `Barrier` for synchronization. Moved timer so as not to time stuff we don't care about.

**Comment by steveluscher** [comment]
Actually, dropped the synchronization in this PR in favour of #3960. Must land that first before this.

**Review by t-nelson — COMMENTED** [review]
r+ #3960 landing first

**Review by alessandrod — APPROVED** [review]

**Review by t-nelson — APPROVED** [review]

**Pull request merged**